### PR TITLE
Refactor to show sample structure

### PR DESCRIPTION
**Pull request by schlessera**
@fklein-lu , @grappler :

I was not really sure how best to leave comments here on how I think you should change up the structure, so I just did a quick refactor to directly show you what I mean.

The basic flow has been simplified (each `Runner` gets a `Theme`, and then adds properties or issues to that theme), while I added an issue management that can be evaluated within the `ReportGenerator`.

I'll let you examine the code at your own leisure. Don't hesitate to ask me anything, if stuff is not clear.

### CODE REVIEW

**Comment by fklein-lu** [comment]
Thanks for all the work the you have been putting into this. I really like the approach that you took, and for this is a good way to move forward.

@grappler what do you think?

**Comment by schlessera** [comment]
Further observations:

- For the different data objects, you know much better what is needed than I do, so these might need to be further defined/extended/changed according to what you think is needed for the entirety of your use cases.
- I only added this for the `Message` so far, but all of these value objects should have a `__toString()` so that they know for themselves how they need to be rendered on the Console. As an example, you might have different types of `Location`s that you need to use. If each of these knows how they need to be rendered as text, the report generator can simply echo the objects themselves and not care about how they should be represented.
- I changed up the bootstrapping process and the command configuration so that you will later be able to just run them through `vendor/bin/theme-check <path>` if you pull in this package as a dependency through Composer.
- It is very easy now to add additional steps into the pipeline. Each step just adheres to the `RunnerInterface` and can then do its thing. I'm not sure right now whether there are any sensible extensions to include, but one never knows... ;)
- The entire setup has now been abstracted into one variable `$pipeline` in the `ThemeCheckCommand.php` file. This could be split out into an overrideable config file.

**Comment by schlessera** [comment]
I quickly went over the report generation and string generation, and the output is already pretty usable now:

screenshot 2017-01-15 08 30 06

File: src/ThemeCheckCommand.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        $stopwatch->start('Theme Check');
+
+        $pipeline = [
+            AnalyserStack::class   => [
+                Analyser\FakeDemoAnalyser::class,
+            ],
+            CheckRunner::class     => [
+                Check\FakeDemoCheck::class,
+            ],
+            ReportGenerator::class => new IOInterfaces($input, $output),
```
Comment (grappler) [inline src/ThemeCheckCommand.php]: This class is not being included. I get the error `Class 'WPTRT\CheckerCli\IOInterfaces' not found`

Comment (schlessera) [inline src/ThemeCheckCommand.php]: Sorry, added the missing class.

**Comment by grappler** [comment]
Thank you @schlessera for your help.

For me the important thing is being able to 
- integrate PHPCS results
- integrated in WordPress plugin or even used in a IDE like PHPCS is today
- return an error code if a theme does not pass the checks that the process does not continue. For example on w.org when a theme is uploaded and does not pass the tests that a new trac ticket is not created.

For PHPCs we can generate a JSON file which we can use to import it into this system. e.g.
https://gist.github.com/grappler/eedfbf2d4c4d007db1be1d9722a7f1cb

For the location I think we need to addition to the line number the column also.

If @fklein-lu agrees we can merge this and then I work on integrating PHPCS to test the code out. If @schlessera would be so kind to do the code review afterwards.

I think this will be for a separate issue but for a first test it would be good to try to add a check that creates an error if the theme tag `custom-header` is used but `add_theme_support( 'custom-header' );` is not defined in the theme.

**Comment by schlessera** [comment]
@grappler:

For PHPCS, if you look through the source, you might find a way to directly use it through code and get back an array (not sure how easy that would be).

Integration should be easy and was already planned for, you just need to provide an alternative bootstrap file, and maybe an alternative to `ThemeCheckCommand.php`.

For the location, I already for having different types of locations. The `FileLocation` is just one example, and you can use multiple different ones as needed.

**Comment by grappler** [comment]
> For PHPCS, if you look through the source, you might find a way to directly use it through code and get back an array (not sure how easy that would be).

From looking at the PHPCS source, there does not seem to be a simple way of getting an error returned as PHPCS expect the output to be either printed to the cli screen or to a file. Using a JSON array would be for a compromise. 

> Integration should be easy and was already planned for, you just need to provide an alternative bootstrap file, and maybe an alternative to `ThemeCheckCommand.php`.

I am not sure I fully understand how it would work. How would the output look like in the end once integrated with PHPCS. From my talk with @fklein-lu last year I understood that this tool was supposed to combine the output of multiple tools together for one report.

**Comment by schlessera** [comment]
For the integration bit, that was referring to this bullet point you mentioned above:

> integrated in WordPress plugin or even used in a IDE like PHPCS is today

**Comment by fklein-lu** [comment]
PHPCs does only seem to offer XML exports. For this software, we can export in whatever format we want.

The integration with WordPress would consist of a light wrapper around the output of both PHPCS, and this tool.

**Comment by grappler** [comment]
> PHPCs does only seem to offer XML exports. For this software, we can export in whatever format we want.

We have JSON export too.

> The integration with WordPress would consist of a light wrapper around the output of both PHPCS, and this tool.

So we would be building another static PHP analyser?

**Comment by fklein-lu** [comment]
We are building a WordPress policy analyser. It means that we don't analyse PHP code, we analyse the correct usage of WordPress APIs. We need to parse PHP to find out how the APIs are used, so it's part of the process.

As far as tools are concerned, I'm open to any suggestions that will achieve this goal.

**Comment by grappler** [comment]
Thanks @fklein-lu I will open a new issue for further discussion. Did you want to merge this?

**Pull request merged**